tests/utils.py

```diff
@@ -8,14 +7,12 @@
 
 @asynccontextmanager
 async def run_server(config: Config, sockets=None):
-    server = Server(config=config)
-    cancel_handle = asyncio.ensure_future(server.serve(sockets=sockets))
-    await asyncio.sleep(0.1)
+    server = Server(config=config, sockets=sockets)
+    await server.start_serving()
     try:
         yield server
     finally:
         await server.shutdown()
-        cancel_handle.cancel()
```

Comment:
I think it's worth pointing this bit out as it suffers from all of the same problems that our test fixtures have suffered from:
  - `asyncio.sleep(0.1)` instead of checking for startup properly.
  - Two handles required to be able to shutdown the service fully.
  - `server.shutdown()` not actually killing the server main loop.
  
I think it's a good demonstration of my motivation for proposing these changes.

Reply:
yep pretty interesting, unfortunate I cant dedicate more time on this, the only real "interrogation" I have it the interaction it may have with --reload  and --workers

Reply:
going to play with this but re-reading it I like it,
@florimondmanca what are your thoughts on this, you extensively played with this a while ago, your input should be very interesting